### PR TITLE
enhancement(snapshot): added support for cross region snapshot copy

### PR DESCRIPTION
**Pull request by BShrivastav**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
=== RUN   TestAccIBMISSnapshotSourceSnapshotCopy
--- PASS: TestAccIBMISSnapshotSourceSnapshotCopy (186.58s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/vpc     188.974s
```

### CODE REVIEW

File: ibm/service/vpc/resource_ibm_is_snapshot.go [hunk 1]
```diff
@@ -83,6 +86,80 @@ func ResourceIBMSnapshot() *schema.Resource {
 				Description:  "Snapshot name",
 			},
 
+			isSnapshotSourceSnapshotCRN: {
```
Comment (deepaksibm) [inline ibm/service/vpc/resource_ibm_is_snapshot.go:89]: sourcesnapshot crn and source volume both are marked as required. But the apispec says they are OneOf.

File: ibm/service/vpc/resource_ibm_is_snapshot.go [hunk 2]
```diff
@@ -83,6 +86,80 @@ func ResourceIBMSnapshot() *schema.Resource {
...
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Source Snapshot CRN",
+			},
+
```
Comment (deepaksibm) [inline ibm/service/vpc/resource_ibm_is_snapshot.go:94]: I see a new field "copies" in the spec.
Also, encryption_key is added to accept key crn.

File: ibm/service/vpc/resource_ibm_is_snapshot.go (hunk 1 above)
Comment (deepaksibm) [reply ibm/service/vpc/resource_ibm_is_snapshot.go:89]: also add acceptance test

File: ibm/service/vpc/data_source_ibm_is_snapshots.go [hunk 3]
```diff
@@ -71,6 +113,74 @@ func DataSourceSnapshots() *schema.Resource {
 							Computed: true,
 						},
 
+						isSnapshotCopies: {
```
Comment (deepaksibm) [inline ibm/service/vpc/data_source_ibm_is_snapshots.go:116]: this should be added in snapshot data source as well

Comment (deepaksibm) [reply ibm/service/vpc/data_source_ibm_is_snapshots.go:116]: add document changes for all new arguments added

File: ibm/service/vpc/resource_ibm_is_snapshot_test.go [hunk 4]
```diff
@@ -268,3 +300,44 @@ func testAccCheckIBMISSnapshotConfigUsertag(vpcname, subnetname, sshname, public
...
+		keys = [ibm_is_ssh_key.testacc_sshkey.id]
+		network_interfaces {
+		  subnet = ibm_is_subnet.testacc_subnet.id
+		  name   = "eth1"
+		}
+	  }
+	resource "ibm_is_snapshot" "testacc_snapshot" {
+		name 			= "%s"
+		source_volume 	= ibm_is_instance.testacc_instance.volume_attachments[0].volume_id
+		source_snapshot_crn = "%s"
```
Comment (deepaksibm) [inline ibm/service/vpc/resource_ibm_is_snapshot_test.go:340]: can't have both here as they are mutually exclusive

File: ibm/service/vpc/resource_ibm_is_snapshot_test.go [hunk 5]
```diff
@@ -104,6 +104,38 @@ ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCKVmnMOlHKcZK8tpt3MP1lqOLAcqcJzhsvJcjscgVE
 	})
 }
 
+// func TestAccIBMISSnapshotSourceSnapshot(t *testing.T) {
```
Comment (deepaksibm) [inline ibm/service/vpc/resource_ibm_is_snapshot_test.go:107]: are going to remove or modify this ?

File: website/docs/d/is_snapshots.html.markdown [hunk 6]
```diff
@@ -61,7 +61,22 @@ In addition to all argument reference list, you can access the following attribu
...
+      - `more_info` - (String) Link to documentation about deleted resources.
+    - `href` - (String) The URL for the copied snapshot.
+    - `id` - (String) The unique identifier for the copied snapshot.
+    - `name` - (String) The name for the copied snapshot. The name is unique across all snapshots in the copied snapshot's native region.
+    - `remote` - (List) If present, this property indicates the referenced resource is remote to this region,and identifies the native region.
+      Nested scheme for `remote`:
+      - `href` - (String) The URL for this region.
+      - `name` - (String) The globally unique name for this region.
+    - `resource_type` - (String) The resource type.
+    - `crn` - (String) The CRN for this snapshot.
```
Comment (deepaksibm) [inline website/docs/d/is_snapshots.html.markdown:79]: crn should move to outer level ?

File: website/docs/r/is_snapshot.html.markdown [hunk 7]
```diff
@@ -87,7 +87,11 @@ Review the argument references that you can specify for your resource.
   **&#x2022;** `access_tags` must be in the format `key:value`.
 - `name` - (Optional, String) The name of the snapshot.
 - `resource_group` - (Optional, Forces new resource, String) The resource group ID where the snapshot is to be created
-- `source_volume` - (Required, Forces new resource, String) The unique identifier for the volume for which snapshot is to be created. 
+- `source_volume` - (Optional, Forces new resource, String) The unique identifier for the volume for which snapshot is to be created.
+- `source_snapshot_crn` - (Optional, String) The CRN for source snapshot.
```
Comment (deepaksibm) [inline website/docs/r/is_snapshot.html.markdown:91]: this also forces new resource ?

File: website/docs/r/is_snapshot.html.markdown [hunk 8]
```diff
@@ -115,6 +119,21 @@ In addition to all argument reference list, you can access the following attribu
 - `resource_type` - (String) The resource type.
 - `size` - (Integer) The size of this snapshot rounded up to the next gigabyte.
 - `source_image` - (String) If present, the unique identifier for the image from which the data on this volume was most directly provisioned.
+- `source_snapshot` - (String) If present, the source snapshot this snapshot was created from.
```
Comment (deepaksibm) [inline website/docs/r/is_snapshot.html.markdown:122]: also add 'copies' in the doc ?

File: ibm/service/vpc/resource_ibm_is_snapshot.go [hunk 9]
```diff
@@ -83,6 +86,81 @@ func ResourceIBMSnapshot() *schema.Resource {
 				Description:  "Snapshot name",
 			},
 
+			isSnapshotSourceSnapshotCRN: {
+				Type: schema.TypeString,
```
Comment (deepaksibm) [inline ibm/service/vpc/resource_ibm_is_snapshot.go:90]: this also forces new resource ?

File: ibm/service/vpc/resource_ibm_is_snapshot.go (hunk 2 above)
Comment (deepaksibm) [reply ibm/service/vpc/resource_ibm_is_snapshot.go:95]: couldn't see isSnapshotEncryptionKey defined in schema that takes input

File: website/docs/r/is_snapshot.html.markdown [hunk 10]
```diff
@@ -87,7 +87,11 @@ Review the argument references that you can specify for your resource.
   **&#x2022;** `access_tags` must be in the format `key:value`.
 - `name` - (Optional, String) The name of the snapshot.
 - `resource_group` - (Optional, Forces new resource, String) The resource group ID where the snapshot is to be created
-- `source_volume` - (Required, Forces new resource, String) The unique identifier for the volume for which snapshot is to be created.
```
Comment (deepaksibm) [inline website/docs/r/is_snapshot.html.markdown:90]: add encryption_key in the arguments

File: ibm/service/vpc/resource_ibm_is_snapshot.go [hunk 11]
```diff
@@ -264,22 +410,61 @@ func resourceIBMISSnapshotCreate(d *schema.ResourceData, meta interface{}) error
...
+			CRN: &encryptionKeyString,
+		}
 	}
-	if grp, ok := d.GetOk(isVPCResourceGroup); ok {
-		rg := grp.(string)
-		snapshotprototypeoptions.ResourceGroup = &vpcv1.ResourceGroupIdentity{
-			ID: &rg,
+
+	// source snapshot
+	if sourceSnaphot, okss := d.GetOk(isSnapshotSourceSnapshotCRN); okss {
```
Comment (deepaksibm) [inline ibm/service/vpc/resource_ibm_is_snapshot.go:447]: add this in the else case as both are mutually exclusive

File: ibm/service/vpc/resource_ibm_is_snapshot_test.go (hunk 5 above)
Comment (BShrivastav) [reply ibm/service/vpc/resource_ibm_is_snapshot_test.go:107]: keeping this.

File: ibm/service/vpc/resource_ibm_is_snapshot.go (hunk 9 above)
Comment (BShrivastav) [reply ibm/service/vpc/resource_ibm_is_snapshot.go:90]: ?

File: website/docs/r/is_snapshot.html.markdown (hunk 7 above)
Comment (BShrivastav) [reply website/docs/r/is_snapshot.html.markdown:105]: removed.

File: website/docs/r/is_snapshot.html.markdown [hunk 12]
```diff
@@ -99,9 +99,14 @@ Review the argument references that you can specify for your resource.
   **&#x2022;** You must have the access listed in the [Granting users access to tag resources](https://cloud.ibm.com/docs/account?topic=account-access) for `access_tags`</br>
   **&#x2022;** `access_tags` must be in the format `key:value`.
 - `clones` - (Optional, List) The list of zones to create a clone of this snapshot.
+- `encryption_key` - (String) A reference CRN to the root key used to wrap the data encryption key for the source snapshot.
 - `name` - (Optional, String) The name of the snapshot.
 - `resource_group` - (Optional, Forces new resource, String) The resource group ID where the snapshot is to be created
-- `source_volume` - (Required, Forces new resource, String) The unique identifier for the volume for which snapshot is to be created. 
+- `source_volume` - (Optional, Forces new resource, String) The unique identifier for the volume for which snapshot is to be created.
+- `source_snapshot_crn` - (Optional, Forces new resource, String) The CRN for source snapshot.
```
Comment (astha-jain) [inline website/docs/r/is_snapshot.html.markdown:106]: Please add the example for cross region i.e. source_snapshot_crn

File: ibm/service/vpc/resource_ibm_is_snapshot_test.go [hunk 13]
```diff
@@ -343,3 +373,43 @@ func testAccCheckIBMISSnapshotCloneConfig(vpcname, subnetname, sshname, publicKe
...
+		zone = "%s"
+		keys = [ibm_is_ssh_key.testacc_sshkey.id]
+		network_interfaces {
+		  subnet = ibm_is_subnet.testacc_subnet.id
+		  name   = "eth1"
+		}
+	  }
+	resource "ibm_is_snapshot" "testacc_snapshot" {
+		name 			= "%s"
+		source_snapshot = "%s"
```
Comment (astha-jain) [inline ibm/service/vpc/resource_ibm_is_snapshot_test.go:412]: source_snapshot or source_snapshot_crn?

**Comment by deepaksibm** [comment]
looks good

**Comment by ujjwal-ibm** [comment]
moving to upstream

**Pull request closed without merging**